### ISSUE TITLE
Planning to work on an environment for Super Smash Bros

### ISSUE THREAD

**Issue by bcoopers** [bzier/gym-mupen64plus#59]

**Comment by bzier** [comment]
@bcoopers Thanks for expressing an interest in taking this on (and for your early contributions already). I have not done any work yet to support additional games besides Mario Kart. As you implement this, feel free to reach out for suggestions or discussion as necessary. I attempted to build the system with multiple games in mind, so my hope is that it should be somewhat straight-forward to add another game. I'm all for you (or anyone) taking this and running with it. However, if you find that you are needing to make significant changes to the core system, please coordinate those with me. I'm pretty open to new or different ideas/approaches to things, but want to make sure that I maintain compatibility with the Mario Kart game, as well as a consistent design and organization in the repository. Hopefully that doesn't sound too authoritarian, but I'd rather just coordinate big changes up-front rather than have you re-write stuff for a month and have me just reject your pull request ;)

I'm more than happy to pitch in on this as you get into it, so let me know if there's anything you'd like my support on. Thanks again for diving head-first into this. Welcome to the party.

**Comment by bcoopers** [comment]
Hey, thanks. I think I've got most stuff set up (menu navigation, character selection, level selection). The main thing I need now is to read in the health on the screen from the pixels so I can fully write the reward function, I'll prolly tackle that soon.

The setup required almost no changes to the core, certainly nothing backwards incompatible- I think all I did was add some new buttons for pressing the joystick all the way left/right/up/down which makes navigating the character select menu much faster (I specifically added new ones instead of overwriting the old ones for backwards compatibility purposes- let me know if you think this was the correct way to do this. I'm not sure where exactly you use the default controls for the Mario Kart, if it's just for menu navigation then oit might be alright to overwrite).

If you wanna take a look, https://github.com/bcoopers/gym-mupen64plus/tree/smash

Thanks!!

**Comment by bzier** [comment]
Very cool... Based on the title of the issue, I wasn't expecting you to be this far along already. I skimmed through the code briefly and it looks like a great start. I'll try to pull it down and check it out at some point today or tomorrow. Plenty of stuff to look through in the smash env (as would be expected).

BTW, I'm a little surprised the new joystick button presses worked for you and that you saw a difference. I was under the impression that the joystick values were constrained between [-80, 80]. Looks like you pushed them out to [-127, 128]. If it works, that's great, I'm just a little surprised. If I can confirm that they are valid values, I may suggest just replacing the existing left/right/up/down since my intent was to do the min/max anyway.

Great work. I look forward to seeing the final product :+1:

**Comment by bcoopers** [comment]
Your Mario Kart implementation was very clear, so it didn't take me very long to adapt it.

Nintendo 64 uses 32 bits to communicate controller state- the first 16 are the buttons, and the second 16 represent the joystick X and Y: https://gaming.stackexchange.com/questions/9264/how-are-the-many-signals-of-a-n64-or-wii-joystick-sent-to-the-console-by-only-2

You can also see here that Mupen64 uses 8 bits for joystick X and Y: https://github.com/libretro/parallel-n64/blob/master/mupen64plus-core/src/api/m64p_plugin.h#L155

The only thing in question then is the input bot, which I briefly looked at and didn't see anything that would max it out to 80. And from my test on Smash, using 127/-128 worked fine, and did go faster that +-80.

**Comment by bzier** [comment]
Excellent, I'm glad to hear that it was clear. Always good to get confirmation from another developer with different perspectives.

Looks like you're right about the joystick values. I was trying to remember why I had it in my head that it was 80. I found [this](https://github.com/mupen64plus/mupen64plus-input-sdl/blob/4aff87c99c13bdcbbf4b9db18dff90f1d9bc2502/src/plugin.c#L280) in the SDL input plugin, which seems to indicate that it is a signed char, which would imply [-128,127], but they also say it is within the range [-80,80] with the calculation `(32768 / 409 = ~80.1)` beside it. I'm not sure where they are getting those numbers though...

I'll try to experiment with it and see how it affects Mario Kart by stretching out those values... perhaps I've been preventing really sharp turns :)

From the input bot, we're actually parsing each of the values as integers, which could potentially lead to problems (if nothing else, overflow somewhere downstream). I created an issue for that [here](https://github.com/kevinhughes27/mupen64plus-input-bot/issues/6). Obviously hasn't been problematic to this point, but better to handle it properly.

**Comment by bzier** [comment]
Hey @bcoopers, what version of the SSB ROM are you using? I pulled your code and attempted to run, but it is getting caught up in the menu navigation. It looks like it is trying to configure game options (turn off items, load options menu, load items menu). However, my game does not seem to have the items menu you are attempting to navigate. I tried two different versions of the ROM, both of which claim to be USA versions, looked the same and did not work.

The MD5 sums of the two ROMs I tried are:
```
3c4e65d9b7bf55338496108f04da7f41
f7c52568a31aadf26e14dc2b6416b2ed
```

Here's the screenshot of the VS Options menu I have:

![image](https://user-images.githubusercontent.com/714528/40283420-ba6deca4-5c32-11e8-977d-c36f78b4c0bb.png)

**Comment by bzier** [comment]
That likely explains the assumption that the CP badge will be on Yoshi every time too. As I was playing around with it, it seemed that the CP position was random each time. Does the save file account for that as well?

I think that makes sense to include the save file, especially if the setup will depend on that being available. Then I think it would good to allow the individual game environments to pass the `--savestate (filepath)` argument to the emulator to pull in their own individual save states and not affect a user's saves outside of executing in the gym environment. To that end, it may also be a good idea to specify the `--nosaveoptions` switch to prevent the settings we're using (e.g. input bot plugin, dummy audio plugin) from affecting user config as well. Or alternatively, specify our own `mupen64plus.cfg` file  with the `--configdir (dir)` switch.

**Comment by bzier** [comment]
An alternative to using the save game (at least for this use case) would be to utilize the game cheats. It looks like SSB has a cheat for `Have All Characters` and `Enable Item Switch Menu`, although at first glance I'm not seeing one listed for the locked level.

You can take a look with `mupen64plus --cheats list smash.n64`

**Comment by bcoopers** [comment]
Yeah I tried the cheats initially, unfortunately they didn't seem to work, the game wouldn't load with the necessary cheats enabled.

I haven't tested yet if the CP position is dependent on the save file. It is definitely dependent on how you navigate the menus prior to selecting the CP. I'll test out whether it depends on the save file soon.

**Comment by bcoopers** [comment]
It's turning out to be a bit more difficult to read the health than I expected. I wrote some code to try to match digits but it fails a decently large percentage of the time.

I've also started exploring potentially reading it from the game's RAM in a Mupen64Plus plugin, though I'm having trouble figuring out how exactly one would do that. I found some memory debugging functions here: https://github.com/mupen64plus/mupen64plus-core/blob/master/src/api/m64p_debugger.h#L137 but I'm having trouble figuring out how to call them from a plugin.

**Comment by bzier** [comment]
I started exploring the RAM idea a couple weeks ago. I bumped into the same hurdles. I'm not quite sure how, first of all, to read the memory in a plugin; and secondly, I'm not sure how to fit it architecturally into the gym wrapper. That would greatly simplify things if we can get that wired up and working though. I'm open to ideas if you've got them. I'll certainly post back if I come up with anything either. Now I have a little more motivation than before with this other environment getting built, so maybe I'll figure something out.

What were you using to match the digits? Was it some kind of CV library, or just custom pixel sampling code, or something else?

**Comment by bcoopers** [comment]
Just got something working quite well, with OpenCV. It's committed to the branch. On my sample of 111 screenshots it gets ~90% correct, and none of them wrong (the remaining 10% return an error message). Which is fortunate, because I made no progress on reading from RAM in a plugin- I think the emulator needs to be built in debugging mode, and I wasn't even able to get Mupen64Plus to compile from source...

**Comment by bzier** [comment]
Glad to hear you're making progress. I'll take a look at what you've got. Before having looked at the code, which very well may answer my question, what happens in the error case? Does it just assume the health hasn't changed and use the previously known value?


Thanks for keeping me up to date on your progress, I appreciate it.

**Comment by bzier** [comment]
Looks like you built out the parser, but haven't plugged it in to the game yet. I didn't study every line, but it looks promising. Haven't done much with OpenCV before so I'll need to spend a little more than 10 min to get my head around all that's happening there.

**Comment by bcoopers** [comment]
Just plugged it into the game. Seems to be working very well. It lags behind a couple of frames with the current implementation, I might work to relax it a bit.

**Comment by bcoopers** [comment]
Submitted a PR. I think pretty much everything is working at this point.

I wasn't able to get --savestate working, it didn't like the save file for some reason. I think maybe it was expected a saved state instead of a saved game file? Anyway I included the saved game file and instructions of how to copy it if you don't have everything unlocked.

I also modified the save file a bit and still deterministically got Yoshi every time, so I think it's just based on your inputs prior to character selection as to which CP gets selected, and not the save file (so long as everything is unlocked of course)

**Comment by bzier** [comment]
I think I may have a solution to the save file problem. I'm still working through some of the details, but we may be able to simplify that, rather than relying on people having the appropriate conditions met. I'll fill you in with the details once I've proven it out and have a slightly better idea of how I'm planning to integrate it.

**Comment by bzier** [comment]
Added #65 with my thoughts around the save files.

**Comment by bzier** [comment]
Closed by #63
